src/settings/settingsdialog.cpp

```diff
@@ -397,6 +405,57 @@ void SettingsDialog::speakYandexTestText()
     ui->yandexPlayerButtons->speak(ui->yandexTestSpeechEdit->text(), m_yandexTranslator->sourceLanguage(), QOnlineTranslator::Yandex);
 }
 
+void SettingsDialog::onGoogleLanguageSelectionChanged(int languageIndex)
+{
+    auto const configuredLang = ui->googleLanguageComboBox->itemData(languageIndex).value<QOnlineTranslator::Language>();
+    auto const originalRegion = ui->googlePlayerButtons->regions(QOnlineTranslator::Google)[configuredLang]; // It will be lost after googleRegionComboBox is changed if not stored here
+
+    ui->googleRegionComboBox->clear();
+
+    ui->googleRegionComboBox->addItem(tr("Default region"), QLocale::AnyCountry);
+    for (auto validRegion : QOnlineTts::validRegions(configuredLang)) {
+        if (validRegion == QLocale::China)
+            ui->googleRegionComboBox->addItem(tr("Mandarin (China)"), QLocale::China); // for now there's only one Chinese dialect supported
+        else
+            ui->googleRegionComboBox->addItem(QLocale::countryToString(validRegion), validRegion);
+    }
+
+    if (originalRegion == QLocale::China)
+        ui->googleRegionComboBox->setCurrentText(tr("Mandarin (China)"));
+    else
+        ui->googleRegionComboBox->setCurrentText(QLocale::countryToString(originalRegion));
+}
+
+void SettingsDialog::saveGoogleEngineRegion(int region)
+{
+    auto lang = ui->googleLanguageComboBox->currentData().value<QOnlineTranslator::Language>();
+
+    auto regionSettings = ui->googlePlayerButtons->regions(QOnlineTranslator::Google);
+    regionSettings[lang] = ui->googleRegionComboBox->itemData(region).value<QLocale::Country>();
+
+    ui->googlePlayerButtons->setRegions(QOnlineTranslator::Google, regionSettings);
+}
+
+void SettingsDialog::detectGoogleTextLanguage()
+{
+    if (ui->googleTestSpeechEdit->text().isEmpty()) {
+        QMessageBox::information(this, tr("Nothing to play"), tr("Playback text is empty"));
+        return;
+    }
+
+    m_googleTranslator->detectLanguage(ui->googleTestSpeechEdit->text(), QOnlineTranslator::Google);
+}
+
+void SettingsDialog::speakGoogleTestText()
+{
+    if (m_googleTranslator->error() != QOnlineTranslator::NoError) {
+        QMessageBox::critical(this, tr("Unable to detect language"), m_googleTranslator->errorString());
+        return;
+    }
+
+    ui->googlePlayerButtons->speak(ui->googleTestSpeechEdit->text(), m_googleTranslator->sourceLanguage(), QOnlineTranslator::Google);
+}
```

Comment:
Looks like this two functions are the same as `detectYandexTextLanguage` and `speakYandexTestText`?
What if we add two helpers that accept `QOnlineTranslator` and `QOnlineTranslator::Engine`? And then call them from these slots. Put them in private section as static functions.

Reply:
Seems the helper functions can't be made static as they need access to this->ui

Reply:
Right, they should be non-static then.

Reply:
I think I have a better idea: two translators are excessive and they are exactly the same but in name. After all, `QOnlineTranslator` stores no engine settings. Only one generic `QOnlineTranslator` instance is needed. What do you think of it?

Reply:
I like the idea!

Reply:
Well, I tried this bright idea and it in fact did no good :(
I'll revert to the original.

Reply:
What was the issue?